Feat: in given out for /custom-direct-quote

deividaspetraitis opened the pull request:
This PR implements in given out for /custom-direct-quote.

Several tests were added to ensure old and new functionality correctness. Some code parts were removed as result to moving towards encapsulated requests. 

Note that test`TestGetCustomQuote_GetCustomDirectQuotesInGivenOut_Mainnet_UOSMOUSDC` requires some improvements as its counterpart `TestGetCustomQuote_GetCustomDirectQuotesInGivenOut_Mainnet` but I am not sure do we want to spent time on that right now.

Note that there are no integration tests either ( old functionality is not covered too ).

Changes includes updating swagger docs.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Enhanced API documentation for the `/router/custom-direct-quote` endpoint, providing detailed parameter descriptions and usage examples.
	- Introduced a new method for retrieving custom quotes based on a specified output token in multi-pool scenarios.
	- Added structured request handling for custom token swap quotes.

- **Bug Fixes**
	- Improved error handling and response clarity for various input validation scenarios.

- **Tests**
	- Expanded test coverage for direct custom quote functionality, including new test cases for exact amount in and out scenarios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

deividaspetraitis committed:
Feat: in given out for /custom-direct-quote

p0mvn commented on domain/mvc/router.go:
```diff
@@ -69,6 +69,9 @@ type RouterUsecase interface {
 	// Otherwise it implements same rules as GetCustomDirectQuote.
 	GetCustomDirectQuoteMultiPool(ctx context.Context, tokenIn sdk.Coin, tokenOutDenom []string, poolIDs []uint64) (domain.Quote, error)
+	// GetCustomDirectQuoteMultiPool calculates direct custom quote for given tokenOut and tokenInDenom over given poolID route.
+	// Otherwise it implements same rules as GetCustomDirectQuote.
```
Let's be more specific about what ie means to implement the same rules

deividaspetraitis replied:
I have updated docs to use different terminology to avoid duplicating `GetCustomDirectQuote` description for each method that uses it as underlying mechanism. Does it make sense?

p0mvn commented on router/delivery/http/router_handler.go:
```diff
@@ -150,39 +148,66 @@ func (a *RouterHandler) GetOptimalQuote(c echo.Context) (err error) {
+		tokenOutDenom []string
+	)
+
+	if req.SwapMethod() == domain.TokenSwapMethodExactIn {
+		tokenIn, tokenOutDenom = req.TokenIn, req.TokenOutDenom
+	} else {
+		tokenIn, tokenOutDenom = req.TokenOut, req.TokenInDenom
+	}
+
+	var quote domain.Quote
+	if req.SwapMethod() == domain.TokenSwapMethodExactIn {
+		quote, err = a.RUsecase.GetCustomDirectQuoteMultiPool(ctx, *tokenIn, tokenOutDenom, req.PoolID)
+	} else {
+		quote, err = a.RUsecase.GetCustomDirectQuoteMultiPoolInGivenOut(ctx, *tokenIn, tokenOutDenom, req.PoolID)
+	}
```
Do we have to do this check twice or can we group it under a single if/else?

deividaspetraitis replied:
Since I have updated to include `humanDenoms` functionality, I think now grouping to in a single branch would add more readability issues and duplicated code. Please let me know what do you think. :pray:

p0mvn commented on router/types/get_direct_custom_quote_request.go:
```diff
@@ -0,0 +1,130 @@
+
+// UnmarshalHTTPRequest unmarshals the HTTP request to GetDirectCustomQuoteRequest.
+// It returns an error if the request is invalid.
+// NOTE: Currently method for some cases returns an error, while for others
+// it returns a response error. This is not consistent and should be fixed.
+func (r *GetDirectCustomQuoteRequest) UnmarshalHTTPRequest(c echo.Context) error {
```
Not sure I'm following the consistency take here

deividaspetraitis replied:
I am referring that within a method it returns `return c.JSON(domain.GetStatusCode(err), domain.ResponseError{Message: err.Error()})` or for example `return ErrTokenInNotValid`, which is inconsistent and sometimes lead to some confusion. I have ported such behaviour from original implementation to avoid different behaviour for the clients.

p0mvn replied:
I see - is there a reason why we cannot fix it right now by always returning an error and then letting the caller handle this as needed?

deividaspetraitis replied:
I think it could be done right now, but at the same time I feel like we already spent some reasonable amount of time improving existing code base while working on this task. From my experience constantly iterating and improving bit by bit allows move faster and hepls to avoid stucking for too long on one code block. Let me know if you think I should still address this particular improvement under this PR. :pray:

p0mvn replied:
No problems - I understand this stream of work has been ongoing for some time with many iterations. It should be pretty quick - I will do it in a follow-up PR

p0mvn commented on router/types/get_direct_custom_quote_request.go:
```diff
@@ -0,0 +1,130 @@
+}
+
+// Validate validates the GetQuoteRequest.
+func (r *GetDirectCustomQuoteRequest) Validate() error {
```
Can we generate tests for this with ChatGPT?

deividaspetraitis replied:
This method is [covered](https://github.com/osmosis-labs/sqs/pull/420/files#diff-92639a1dec5495d469549297be5307e13d23fd17e4407b2866dbe3b9ce9b3608R198) with tests, or you want actually just feed it to ChatGPT?

p0mvn reviewed (commented):
Nice work!

Could you please add am integration test and ensure it passes as well as previous "in given out" tests? Then, we can merge this 🙏

p0mvn reviewed (commented):
This query seems to be failing:
```
http://206.81.19.9:9092/router/custom-direct-quote?tokenOut=1000000ibc%2F27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2&tokenInDenom=uosmo%2Cuion&poolID=1%2C2
```

The failure occurs in `PrepareResult`.

Could we create a functional mainnet test off this example?

p0mvn commented on router/delivery/http/router_handler.go:
```diff
@@ -150,39 +148,66 @@ func (a *RouterHandler) GetOptimalQuote(c echo.Context) (err error) {
-// @Param  tokenIn         query  string  true  "String representation of the sdk.Coin for the token in."                   example(5OSMO)
-// @Param  tokenOutDenom   query  string  true  "String representing the list of the token denom out separated by comma."   example(ATOM,USDC)
-// @Param  poolID          query  string  true  "String representing list of the pool ID."                                  example(1,2,3)
+// @Param  tokenIn         query  string  false  "String representation of the sdk.Coin denoting the input token for the exact amount in swap method."                       example(1000000uosmo)
+// @Param  tokenOutDenom   query  string  false  "String representing the list of the output token denominations separated by comma for the exact amount in swap method."    example(uion)
+// @Param  tokenOut        query  string  false  "String representation of the sdk.Coin denoting the output token for the exact amount out swap method."                     example(2353uion)
+// @Param  tokenInDenom    query  string  false  "String representing the list of the input token denominations separated by comma for the exact amount out swap method."    example(uosmo)
```
Would you mind adding `humanDenoms` functionality to this endpoint please? I think it would be a great dev UX improvement

deividaspetraitis replied:
I have added `humanDenoms` including relevant tests covering this functionality. :+1:

p0mvn commented on router/delivery/http/router_handler.go:
```diff
@@ -150,39 +148,66 @@ func (a *RouterHandler) GetOptimalQuote(c echo.Context) (err error) {
+		tokenOutDenom []string
+	)
+
+	if req.SwapMethod() == domain.TokenSwapMethodExactIn {
+		tokenIn, tokenOutDenom = req.TokenIn, req.TokenOutDenom
+	} else {
+		tokenIn, tokenOutDenom = req.TokenOut, req.TokenInDenom
+	}
+
+	var quote domain.Quote
+	if req.SwapMethod() == domain.TokenSwapMethodExactIn {
+		quote, err = a.RUsecase.GetCustomDirectQuoteMultiPool(ctx, *tokenIn, tokenOutDenom, req.PoolID)
+	} else {
+		quote, err = a.RUsecase.GetCustomDirectQuoteMultiPoolInGivenOut(ctx, *tokenIn, tokenOutDenom, req.PoolID)
+	}
```
Is it possible to minimize the number of checks/branching that are the same?

deividaspetraitis replied:
Replied in this [comment](https://github.com/osmosis-labs/sqs/pull/420#discussion_r1711624400). Closing this thread to make this PR more managable.

p0mvn reviewed (commented):
Nice work!

I'll be good with approving and merging upon previous comments being addressed and also an integration PR created where we can run all tests with no failures 🙏

deividaspetraitis committed:
Merge remote-tracking branch 'upstream/v25.x' into SQS-303

deividaspetraitis commented:
Will get back to this later, switching for higher priority tasks for now. Stopped at integration [tests](https://github.com/osmosis-labs/sqs/pull/423): 
> If somebody would call /quote and would try to feed results to /get-direct-custom-quote that would not work for a [multi routes](https://github.com/osmosis-labs/sqs/pull/423/files#diff-9d1423fc5abbd62d4ff5f8151c5277b0dbc5af89a2e603f2b0c211835f690a11R546). I am not entirely sure is that valid case at all? I came up with this edge case following [the idea](https://github.com/osmosis-labs/sqs/pull/423/files#diff-f2dc6df9d608942d3fe071e9ccb7084f8cab5d972326e8291c0bb3354207caeeR30) that I could use results from /quote to generate inputs for testing /get-direct-custom-quote  to avoid flakiness in the tests

deividaspetraitis committed:
Feat: in given out for /custom-direct-quote

Add humanDenoms support

deividaspetraitis committed:
Merge remote-tracking branch 'upstream/v25.x' into SQS-303

deividaspetraitis committed:
Feat: in given out for /custom-direct-quote

Update swagger

deividaspetraitis committed:
Feat: in given out for /custom-direct-quote

Update docs

deividaspetraitis committed:
Merge remote-tracking branch 'upstream/v25.x' into SQS-303

deividaspetraitis committed:
Merge remote-tracking branch 'upstream/v25.x' into SQS-303

deividaspetraitis committed:
Feat: in given out for /custom-direct-quote

Add docs for supported route types

deividaspetraitis commented:
> This query seems to be failing:
> 
> ```
> http://206.81.19.9:9092/router/custom-direct-quote?tokenOut=1000000ibc%2F27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2&tokenInDenom=uosmo%2Cuion&poolID=1%2C2
> ```
> 
> The failure occurs in `PrepareResult`.
> 
> Could we create a functional mainnet test off this example?

This one seems was resolved too.

p0mvn reviewed (approved):
LGTM - nice work!

p0mvn committed:
Feat: in given out for /custom-direct-quote (#423)

* Feat: in given out for /custom-direct-quote

e2e tests for exact amount out init

* fix merge error

* WIP

* WIP

* Feat: in given out for /custom-direct-quote

e2e tests for /custom-direct-quote

* Feat: in given out for /custom-direct-quote

Break run_quote_test into smaller individual parts for reusability

* Feat: in given out for /custom-direct-quote

Introduce calculate_expected_base_out_quote_spot_price

* Feat: in given out for /custom-direct-quote

Add synthetic tests for /quote and /custom-direct-quote

* Update tests/route.py

---------

Co-authored-by: Deividas Petraitis <hi@deividaspetraitis.lt>

p0mvn committed:
code rabbit
